### PR TITLE
Pagination 

### PR DESCRIPTION
**Pull request by Redtigercod4**
This PR includes the following:

Adds Pagination using offset & limit within the Sequelise ORM.
Adds functionality to Get All Assets endpoint
Updated and added new tests to cover new functionality

### CODE REVIEW

File: api/src/tests/helpers/testData.ts [hunk 1]
```diff
@@ -10,3 +11,41 @@ export const testAsset: AssetAttributes = {
...
+      serialNumber: "",
+      modelNumber: "",
+      nextAuditDate: new Date(),
+      createdAt: new Date(),
+      updatedAt: new Date(),
+    },
+    {
+      id: 3,
+      assetTag: "testAssetTag2",
+      name: "testAsset2",
```
Comment (SolarFactories) [inline api/src/tests/helpers/testData.ts:43]: W: Have numbers within `name` and `assetTag` match up with `id`, as opposed to being offset by 1.

File: api/src/services/BaseService.ts [hunk 2]
```diff
@@ -22,5 +26,9 @@ export abstract class BaseService<TEntity extends Model>
 
     return data;
   }
+
+  public async findAllPaginated(page: number, pageSize: number) {
+      return await this.paginator.paginate(page, pageSize)
```
Comment (SolarFactories) [inline api/src/services/BaseService.ts:31]: W (Pedantic): Missing semicolon. Should be being picked up by pipeline. Issue to be created.

File: api/src/tests/unit/controllers/AssetController.test.ts [hunk 3]
```diff
@@ -1,35 +1,48 @@
...
 const logger: MockedLogger = Logger as unknown as MockedLogger;
 
-describe('getAllAssets', () => {
+describe("getAllAssets", () => {
   const assetController: AssetController = new AssetController();
   let request: Request;
   let response: Response;
   let next: NextFunction;
-  let findAllMock: jest.MockedFunction<() => Promise<AssetAttributes[]>>;
+  let findAllMock: jest.MockedFunction<
```
Comment (SolarFactories) [inline api/src/tests/unit/controllers/AssetController.test.ts:34]: W (Pedantic): `findAllPaginatedMock`, to reflect the name of the method which is being mocked.

File: api/src/tests/unit/controllers/AssetController.test.ts [hunk 4]
```diff
@@ -38,35 +51,72 @@ describe('getAllAssets', () => {
...
+      ErrorController.NotFoundError("No Assets Found")
+    );
     expectNonFinal(response);
   });
+
+  it.each([
+    [
+      { page: "0", pageSize: "1" },
+      { page: "1", pageSize: "0" },
+    ],
```
Comment (SolarFactories) [inline api/src/tests/unit/controllers/AssetController.test.ts:96]: S: Currently this is only running the test once, with two arguments. Each `object` should be wrapped in it's own array. W: Have the parameters of `page` and `pageSize` print out in the test name.

File: api/src/tests/unit/utils/Paginator.test.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+  error: jest.fn(),
+}));
+
+const MockedModel: jest.Mocked<ModelStatic<any>> =
+  jest.createMockFromModule("sequelize");
+const mockFindAndCountAll = jest.fn();
+MockedModel.findAndCountAll = mockFindAndCountAll;
+
+describe("Paginator Unit Tests", () => {
+  let paginator: Paginator<any>;
```
Comment (SolarFactories) [inline api/src/tests/unit/utils/Paginator.test.ts:17]: C: Replace `any` with `Asset`.

File: api/src/tests/unit/utils/Paginator.test.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  it("returns paginated test data with 3 total records", async () => {
+
+    mockFindAndCountAll.mockResolvedValueOnce({ rows: testPaginationAssets.data, count: testPaginationAssets.totalRecords });
+
+    const page = 1;
+    const pageSize = 10;
+    const result = await paginator.paginate(page, pageSize);
+
+    expect(result.lastPage).toBe(1);
+    expect(result.totalRecords).toBe(3);
```
Comment (SolarFactories) [inline api/src/tests/unit/utils/Paginator.test.ts:36]: W: Make dynamic to be `testPaginationAssets.length`, to be resilient to new test assets being added.

File: api/src/tests/unit/utils/Paginator.test.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+  it("should throw error for invalid page or pageSize", async () => {
+    const page = 0;
+    const pageSize = 10;
+
+    await expect(paginator.paginate(page, pageSize)).rejects.toThrow();
+    expect(mockFindAndCountAll).not.toHaveBeenCalled();
+
+    await expect(paginator.paginate(pageSize, page)).rejects.toThrow();
+    expect(mockFindAndCountAll).not.toHaveBeenCalled();
+  });
```
Comment (SolarFactories) [inline api/src/tests/unit/utils/Paginator.test.ts:54]: W: Use `it.each` to clarify that it is two scenarios being tested, and not duplicated lines within the same test.

Comment (Redtigercod4) [reply api/src/tests/unit/utils/Paginator.test.ts]: Commit: 02dbeb7a0fa4367798ed906384e2bba6522c1763

File: api/src/tests/unit/utils/Paginator.test.ts (hunk 6 above)
Comment (Redtigercod4) [reply api/src/tests/unit/utils/Paginator.test.ts]: Commit: 02dbeb7a0fa4367798ed906384e2bba6522c1763

File: api/src/tests/unit/utils/Paginator.test.ts (hunk 5 above)
Comment (Redtigercod4) [reply api/src/tests/unit/utils/Paginator.test.ts]: Commit: 02dbeb7a0fa4367798ed906384e2bba6522c1763

File: api/src/tests/unit/controllers/AssetController.test.ts (hunk 4 above)
Comment (Redtigercod4) [reply api/src/tests/unit/controllers/AssetController.test.ts]: Commit: 02dbeb7a0fa4367798ed906384e2bba6522c1763

File: api/src/tests/unit/controllers/AssetController.test.ts (hunk 3 above)
Comment (Redtigercod4) [reply api/src/tests/unit/controllers/AssetController.test.ts]: Commit: 02dbeb7a0fa4367798ed906384e2bba6522c1763

File: api/src/services/BaseService.ts (hunk 2 above)
Comment (Redtigercod4) [reply api/src/services/BaseService.ts]: Commit: 02dbeb7a0fa4367798ed906384e2bba6522c1763

File: api/src/tests/helpers/testData.ts (hunk 1 above)
Comment (Redtigercod4) [reply api/src/tests/helpers/testData.ts]: Commit: 02dbeb7a0fa4367798ed906384e2bba6522c1763

**Review by SolarFactories — APPROVED** [review]

**Pull request merged**